### PR TITLE
scripts: add Claude stop-hook done-claim guard (#925)

### LINKED ISSUE #925: dev: Stop-hook done-claim validator — block turn-end without gate evidence (spec + rollout guide)
Follow-up from the 2026-07-07 workflow audit (companion to #921/#922/#923). The audited failure mode: an agent ends its turn claiming "done" without the verification evidence the CLAUDE.md delegation contract requires (executed gates, red→green proof, gate record). Prose demands it; nothing blocks a turn that omits it. A `Stop`/`SubagentStop` hook can — it fires when the model tries to end its turn and can force one more round of work.

This is deliberately spec-first: Stop hooks are the one hook class that can brick a session when buggy (every turn ends through them), so the implementation must follow this guide exactly and roll out gradually.

## How it works (verify against <https://code.claude.com/docs/en/hooks.md> before coding — do NOT code the schema from memory)

- Register under `hooks.Stop` (and `hooks.SubagentStop` if supported — the docs are the authority) in `.claude/settings.json`.
- The hook receives JSON on stdin including `transcript_path` (the session JSONL) and `stop_hook_active` (true when the turn is already a continuation forced by a previous Stop-hook block — **must be honored to prevent infinite loops**).
- To let the turn end: exit 0 with no output. To block: emit the Stop-hook decision JSON (`{"decision": "block", "reason": "<what is missing>"}` — confirm the exact current schema from the docs); the reason is fed back to the model, which continues working.

## Decision logic (implement in a tested script, e.g. `scripts/claude-stop-guard.py` — dev-host tooling, so Python is fine per CLAUDE.md)

1. **Fail-open on any doubt.** Parse error, missing transcript, unreadable JSONL, any unexpected shape → exit 0. This is a guardrail, not a security boundary; a false block is worse than a missed one.
2. **Loop guard first:** `stop_hook_active` true → exit 0 unconditionally.
3. Read only the tail of the transcript (last ~50 entries) — never the whole file.
4. **Trigger condition (all must hold, else exit 0):**
   - The turn touched code: some `Edit`/`Write` tool_use in the turn targeting `src/**` or `tests/**` (path is in the tool input).
   - The final assistant text makes a completion claim: matches a conservative regex like `\b(done|complete[d]?|implemented|fixed|landed|finished|green)\b` (case-insensitive).
   - No verification command ran **after the last code edit**: no `Bash` tool_use whose command contains any of `pytest`, `phpunit`, `phpstan`, `phpcs`, `php -l`, `ruff`, `shellcheck`, `shellspec`, `mypy`.
5. On trigger: block with a reason that quotes the contract, e.g. `"CLAUDE.md THE GATE: code was edited but no gate command ran after the last edit. Run the canonical gates for the touched languages and report expected-vs-actual output — or state explicitly why no gate applies."`
6. **One block per turn maximum** — guaranteed by the `stop_hook_active` check in (2).

## What it must NOT do

- No network, no git mutations, no writes outside stderr/stdout.
- Never block conversational turns (no code edits → always exit 0) — this keeps Q&A, triage-only, and docs sessions untouched.
- Never re-implement the gate itself; it only checks that *some* gate evidence exists. Depth stays in the skills; the hook is a floor.

## Tests (mandatory, per the test mandate — every branch)

Unit-test the decision function on synthetic transcript-tail fixtures (`tests/test_claude_stop_guard.py`): edits+claim+no-gate → block; edits+claim+gate-after-edit → pass; edits+gate-before-last-edit-only → block; no edits → pass; `stop_hook_active` → pass; garbage/truncated JSONL → pass (fail-open); docs-only edits (`*.md`) → pass.

## Rollout

1. Land script + tests, hook **not** registered.
2. Register in `.claude/settings.local.json` (one machine) for ~a week; watch for false blocks; tune the completion-claim regex if needed.
3. Only then move the registration to the shared `.claude/settings.json`.
4. Escape hatch documented in the reason string itself: the model can satisfy the hook by explaining why no gate applies — so a legitimate edge case costs one extra sentence, never a stuck session.

### PR DESCRIPTION
**Pull request by andrebrait**
Fixes #925 — rollout step 1 only: the guard script and its tests land here; the hook is **not** registered in any settings file.

## What

- `scripts/claude-stop-guard.py` — Claude Code `Stop`-hook guard. Blocks a turn from ending when all three hold: the turn edited a non-`.md` file under `src/` or `tests/`, the final assistant text makes a completion claim (`done`/`completed`/`implemented`/`fixed`/`landed`/`finished`/`green`), and no gate command (`pytest`/`phpunit`/`phpstan`/`phpcs`/`php -l`/`ruff`/`shellcheck`/`shellspec`/`mypy`) ran after the last edit. Block output is the documented `{"decision": "block", "reason": ...}` JSON; the reason quotes CLAUDE.md THE GATE and documents the escape hatch (state why no gate applies).
- `tests/test_claude_stop_guard.py` — 23 tests: all 13 coverage-matrix rows (block/allow branches, per-keyword gate recognition, prior-turn gates excluded, `Write` counts as edit, `last_assistant_message` fallback) plus 9 hostile-input rows (garbage/truncated JSONL, binary lines, oversized ~1MB line, malformed `tool_use` entries, missing transcript, non-JSON stdin, subprocess block-path shape, `undone` vs `DONE.` word boundary).
- `scripts/README.md` — one table row for the new script.

## Spec corrections vs the issue (doc-verified this session)

The issue mandated confirming the hook schema from the current docs rather than coding it from memory. Two findings:

- `stop_hook_active` is **absent** from the currently documented Stop-hook stdin schema. The script still honors it if present (forward/backward compat) and adds a second stateless loop guard: the block reason carries a `[stop-guard]` marker, and a marker already present in the transcript tail means we blocked this turn before — allow. No state files (the spec forbids writes).
- `last_assistant_message` **is** documented on Stop stdin and is preferred for the completion-claim check, falling back to the transcript's last assistant text when absent.

Transcript-entry shapes used by the parser and test fixtures were probed against a real live-session transcript, not recalled.

## Safety properties

- Fail-open everywhere: any exception, unparseable payload, or missing transcript exits 0 silently.
- No network, no file writes, no git operations; stdout/stderr only.
- Conversational turns (no code edits) can never block.

## Verification

- Red→green executed (not reasoned): with the script moved aside, the suite fails collection with `FileNotFoundError`; restored, `23 passed`.
- Gates: `pytest` (23 passed; full-suite failures are the pre-existing zstd-less-container ones in `test_pfb_pkg.py`/`test_build_repo_portable.py`, unrelated), `ruff check`, `ruff format --check`, `mypy tests/` — all clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_017kC9yYSBND3KSKMjGz64vZ

---
_Generated by [Claude Code](https://claude.ai/code/session_017kC9yYSBND3KSKMjGz64vZ)_

### CODE REVIEW

File: scripts/claude-stop-guard.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    try:
+        payload = json.load(sys.stdin)
+        if not isinstance(payload, dict):
+            return 0
+        transcript_path = payload.get("transcript_path")
+        tail_lines: list[str] = []
+        if isinstance(transcript_path, str) and transcript_path:
+            with open(transcript_path, encoding="utf-8", errors="replace") as f:
+                lines = f.readlines()
+            tail_lines = lines[-_TAIL_LINES:]
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply scripts/claude-stop-guard.py]: Fixed in 921c1db — `main()` now seeks to `max(0, size - 5 MB)` and reads only the bounded tail (dropping the partial first line when the cut lands mid-line), instead of `readlines()` on the whole file. The same commit also fixes the deeper variant of this the adversarial review found: the old 50-line slice silently dropped the triggering edit on any turn longer than the window, flipping BLOCK to ALLOW.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/claude-stop-guard.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        return None
+
+    entries = _parse_lines(tail_lines)
+    if not entries:
+        return None
+
+    last_user_idx = _last_genuine_user_index(entries)
+    turn = entries[last_user_idx + 1 :] if last_user_idx is not None else entries
+    if not turn:
+        return None
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply scripts/claude-stop-guard.py:165]: Deliberate, and now much narrower after 921c1db. The whole-tail-as-turn fallback is the issue-spec behavior for a missing boundary, and with the tail now 5 MB (not 50 lines) the boundary can only be lost when a single turn exceeds 5 MB of JSONL. In that corner the failure directions are asymmetric: treating the fallback as fail-open would disarm the guard exactly on the longest turns (its primary target), while the current choice costs at most one extra block — whose reason text carries the escape hatch ("state explicitly why no gate applies") and whose `[stop-guard]` marker prevents any repeat block in the same turn. Fail-open here governs parse/IO doubt, not a tail that genuinely shows edit + claim + no gate. Pinned by `test_main_byte_cap_truncating_user_boundary_still_blocks`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by andrebrait** [comment]
## Review audit — PR #945

**Reviews received:** Claude Sonnet 5 adversarial review via the `review-fanout` workflow (3 lenses → dedup → execution-grounded verification; 11 agents; shape chosen because the PR exceeds the 300-line bar at +472/3 files), a focused Sonnet 5 re-review of the review-fix delta (`921c1db`), and GitHub Copilot (2 comments). **CodeRabbit:** skipped under the 5-minute rule — rate-limited with "next review available in 59 minutes" at PR open, still rate-limited (18 min) at merge time; no review content was ever produced. **Snyk:** `error — Code test limit reached` (quota); did not run, ignored per policy (not a finding, not a pass).

### Findings ledger

Fan-out review (8 confirmed / 0 refuted, deduplicated to 6):

1. **blocking** — 50-line tail cut dropped the triggering edit on any turn longer than the window, silently flipping BLOCK→ALLOW (empirically reproduced via `main()`); → fixed@`921c1db` (5 MB byte-tail from EOF; turn boundary decides scope; pinned by `test_main_long_turn_via_subprocess_blocks` red→green)
2. **blocking** — gate keyword substring match credited `ruff` inside `truffle`; → fixed@`921c1db` (`\b`-anchored `_GATE_RE`; pinned by `test_gate_keyword_inside_larger_word_not_credited`)
3. **blocking** — claim fallback walked back past the last assistant entry, so a stale mid-turn "done" could false-block; → fixed@`921c1db` (last entry only; pinned by `test_stale_done_before_final_tooluse_entry_does_not_block`)
4. **blocking** — only 3 of 7 claim-regex alternatives tested; → fixed@`921c1db` (`test_each_claim_keyword_recognized`, all 8 word forms)
5. nitpick — preferred `last_assistant_message` branch untested; → fixed@`921c1db` (`test_payload_claim_text_preferred_over_transcript`, both directions)
6. nitpick — Edit + gate Bash batched in one assistant entry spuriously blocked; → fixed@`921c1db` (`i < after_index`; pinned by `test_same_entry_batched_edit_and_gate_allows`)

Focused re-review of the fix delta (1 blocking / 2 nitpicks):

7. **blocking** — unconditional partial-first-line drop discarded a COMPLETE first line when the byte cut landed exactly after a newline, re-introducing BLOCK→ALLOW at the boundary (reproduced); → fixed@`7e79520` (drop only when the byte before the cut isn't `\n`; pinned by `test_main_byte_cut_on_exact_line_boundary_keeps_first_line` red→green)
8. nitpick — decide()-level long-turn test comment claimed to pin the windowing bug it cannot see; → fixed@`7e79520` (relabeled, names the subprocess test as the true pin)
9. nitpick — case-insensitive gate matching was an undocumented, untested widening; → fixed@`7e79520` (deliberate, consistent with `_CLAIM_RE`; pinned by `test_gate_keyword_case_insensitive`)

Copilot:

10. `readlines()` whole-file memory concern → fixed@`921c1db`; replied on thread
11. missing-user-boundary fallback can false-block → **skipped (works-as-designed)**: issue-spec fallback behavior, reachable only when one turn exceeds 5 MB, costs at most one block carrying the escape hatch, `[stop-guard]` marker prevents repeats; reasoning + pinning test on the thread

**Gates on the final tip (`7e79520`):** `pytest tests/test_claude_stop_guard.py` → 33 passed; `ruff check` / `ruff format --check` / `mypy tests/` clean; full suite failures confined to the pre-existing zstd-less-container ones (`test_pfb_pkg.py` / `test_build_repo_portable.py`), unrelated to this diff. All behavior-changing fixes carry executed red→green proof against the pre-fix code.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_
